pay-api/src/pay_api/services/eft_service.py

```diff
@@ -108,14 +117,341 @@ def process_cfs_refund(self, invoice: InvoiceModel,
+        """Get most recent EFT Credit invoice links associated to a statement and short name."""
+        query = (db.session.query(EFTCreditInvoiceLinkModel)
+                 .distinct(EFTCreditInvoiceLinkModel.invoice_id)
+                 .join(EFTCreditModel, EFTCreditModel.id == EFTCreditInvoiceLinkModel.eft_credit_id)
+                 .join(StatementInvoicesModel,
+                       StatementInvoicesModel.invoice_id == EFTCreditInvoiceLinkModel.invoice_id)
+                 .filter(StatementInvoicesModel.statement_id == statement_id)
+                 .filter(EFTCreditModel.short_name_id == shortname_id)
+                 .filter(EFTCreditInvoiceLinkModel.status_code != EFTCreditInvoiceStatus.CANCELLED.value)
+                 .order_by(EFTCreditInvoiceLinkModel.invoice_id.desc(),
+                           EFTCreditInvoiceLinkModel.created_on.desc(),
+                           EFTCreditInvoiceLinkModel.id.desc())
+                 )
+        return query.all()
+
+    @staticmethod
+    def apply_eft_credit(invoice_id: int,
+                         short_name_id: int,
+                         link_group_id: int,
+                         auto_save: bool = False):
+        """Apply EFT credit and update remaining credit records."""
+        invoice = InvoiceModel.find_by_id(invoice_id)
+        payment_account = PaymentAccountModel.find_by_id(invoice.payment_account_id)
+
+        # Clear any existing pending credit links on this invoice
+        EftService.cancel_payment_action(short_name_id, payment_account.auth_account_id, invoice_id)
+
+        eft_credit_balance = EFTCreditModel.get_eft_credit_balance(short_name_id)
+        invoice_balance = invoice.total - (invoice.paid or 0)
+
+        if eft_credit_balance < invoice_balance:
+            return
+
+        eft_credits = EftService._get_eft_credits(short_name_id)
+        for eft_credit in eft_credits:
+            credit_invoice_link = EFTCreditInvoiceLinkModel(
+                eft_credit_id=eft_credit.id,
+                status_code=EFTCreditInvoiceStatus.PENDING.value,
+                invoice_id=invoice.id,
+                link_group_id=link_group_id)
+
+            if eft_credit.remaining_amount >= invoice_balance:
+                # Credit covers the full invoice balance
+                credit_invoice_link.amount = invoice_balance
+                credit_invoice_link.save_or_add(auto_save)
+                eft_credit.remaining_amount -= invoice_balance
+                eft_credit.save_or_add(auto_save)
+                break
+
+            # Credit covers partial invoice balance
+            invoice_balance -= eft_credit.remaining_amount
+            credit_invoice_link.amount = eft_credit.remaining_amount
+            credit_invoice_link.save_or_add(auto_save)
+            eft_credit.remaining_amount = 0
+            eft_credit.save_or_add(auto_save)
+
+    @staticmethod
+    def _get_eft_credits(short_name_id: int) -> List[EFTCreditModel]:
```

Comment:
This one strictly querying against the EFT Credit Model, a very simple query. Might be cleaner to just move this one to the model.

Reply:
Should be updated